mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -0,0 +1,422 @@
+            cache_key,
+            part_queue_producer,
+        }
+    }
+
+    async fn get_from_cache(&self, range: RequestRange) {
+        let key = &self.cache_key.s3_key;
+        let block_size = self.cache.block_size();
+        let block_range = self.block_indices_for_byte_range(&range);
+
+        // TODO: consider starting GetObject requests pre-emptively if cache blocks are missing
+        for block_index in block_range.clone() {
+            match self.cache.get_block(&self.cache_key, block_index) {
+                Ok(Some(block)) => {
+                    trace!(?key, ?range, block_index, "cache hit");
+                    let part = self.make_part(block, block_index, &range);
+                    metrics::counter!("cache.total_bytes", part.len() as u64, "type" => "read");
+                    self.part_queue_producer.push(Ok(part));
+                }
+                Ok(None) => {
+                    trace!(?key, ?range, block_index, "cache miss - no data for block");
+                    return self
+                        .get_from_client(range.trim_start(block_index * block_size), block_index..block_range.end)
+                        .await;
+                }
+                Err(error) => {
+                    trace!(?key, ?range, block_index, ?error, "error reading block from cache");
+                    return self
+                        .get_from_client(range.trim_start(block_index * block_size), block_index..block_range.end)
+                        .await;
+                }
+            }
+        }
+    }
+
+    async fn get_from_client(&self, range: RequestRange, block_range: Range<u64>) {
+        let key = &self.cache_key.s3_key;
+        let block_size = self.cache.block_size();
+        assert!(block_size > 0);
+
+        let block_aligned_byte_range =
+            (block_range.start * block_size)..(block_range.end * block_size).min(range.object_size() as u64);
+
+        trace!(
+            ?key,
+            range =? block_aligned_byte_range,
+            original_range =? range,
+            "fetching data from client"
+        );
+        let get_object_result = match self
+            .client
+            .get_object(
+                &self.bucket,
+                key,
+                Some(block_aligned_byte_range),
+                Some(self.cache_key.etag.clone()),
+            )
+            .await
+        {
+            Ok(get_object_result) => get_object_result,
+            Err(e) => {
+                error!(key, error=?e, "GetObject request failed");
+                self.part_queue_producer
+                    .push(Err(PrefetchReadError::GetRequestFailed(e)));
+                return;
+            }
+        };
+
+        pin_mut!(get_object_result);
+        let mut block_index = block_range.start;
+        let mut buffer = ChecksummedBytes::default();
+        loop {
+            match get_object_result.next().await {
+                Some(Ok((offset, body))) => {
```

Comment:
For paranoia, I think we want to keep the `offset` around and do some arithmetic with it to make sure it's always in sync with the `block_index`. Right now I can't see anything stopping us from flipping a bit in `block_index` and then returning the wrong bytes to a future read. Keeping `offset` around would give us a second source of truth for location (one that comes from S3 rather than being computed from `block_index`) that we can plumb all the way down into the cache -- the cache blocks should record their byte offset.

Reply:
Added the check but not stored the byte offset in the cache blocks.

Reply:
Will complete in a separate change